File: frontend/src/api/orgs.ts
import { request } from './request'

export interface PageResult<T> {
  items: T[]
  total: number
  page?: number
  page_size?: number
}

export interface School {
  id: number
  name: string
  code: string | null
  address: string | null
  contact: string | null
  phone: string | null
  description: string | null
  created_at: string
}

export interface Grade {
  id: number
  school_id: number
  school_name: string | null
  name: string
  level: number
  class_count: number
}

export interface Clazz {
  id: number
  grade_id: number
  grade_name: string | null
  school_id: number | null
  name: string
  head_teacher_id: number | null
  head_teacher_name: string | null
  student_count: number
}

export interface Student {
  id: number
  student_no: string
  name: string
  gender: string
  birth_date: string | null
  avatar: string | null
  school_id: number
  grade_id: number
  class_id: number
  class_name: string | null
  parent_name: string | null
  parent_phone: string | null
  enrollment_date: string | null
  note: string | null
}

export interface Teacher {
  id: number
  user_id: number | null
  username: string | null
  teacher_no: string
  name: string
  gender: string
  school_id: number
  phone: string | null
  email: string | null
  title: string | null
  avatar: string | null
}

export interface Subject {
  id: number
  name: string
  code: string | null
  sort_order: number
}

export interface TeacherClassSubject {
  id: number
  teacher_id: number
  teacher_name: string | null
  class_id: number
  class_name: string | null
  grade_id: number | null
  grade_name: string | null
  subject_id: number
  subject_name: string | null
}

const list = <T>(url: string, params?: any) =>
  request<PageResult<T>>({ url, method: 'get', params })

// ---- 学校 ----
export const listSchools = (params?: any) => list<School>('/orgs/schools', params)
export const createSchool = (data: any) => request<School>({ url: '/orgs/schools', method: 'post', data })
export const updateSchool = (id: number, data: any) =>
  request<School>({ url: `/orgs/schools/${id}`, method: 'put', data })
export const deleteSchool = (id: number) =>
  request<void>({ url: `/orgs/schools/${id}`, method: 'delete' })

// ---- 年级 ----
export const listGrades = (params?: any) => list<Grade>('/orgs/grades', params)
export const createGrade = (data: any) => request<Grade>({ url: '/orgs/grades', method: 'post', data })
export const updateGrade = (id: number, data: any) =>
  request<Grade>({ url: `/orgs/grades/${id}`, method: 'put', data })
export const deleteGrade = (id: number) =>
  request<void>({ url: `/orgs/grades/${id}`, method: 'delete' })

// ---- 班级 ----
export const listClasses = (params?: any) => list<Clazz>('/orgs/classes', params)
export const createClass = (data: any) => request<Clazz>({ url: '/orgs/classes', method: 'post', data })
export const updateClass = (id: number, data: any) =>
  request<Clazz>({ url: `/orgs/classes/${id}`, method: 'put', data })
export const deleteClass = (id: number) =>
  request<void>({ url: `/orgs/classes/${id}`, method: 'delete' })

// ---- 学生 ----
export const listStudents = (params?: any) =>
  request<PageResult<Student>>({ url: '/orgs/students', method: 'get', params })
export const createStudent = (data: any) => request<Student>({ url: '/orgs/students', method: 'post', data })
export const updateStudent = (id: number, data: any) =>
  request<Student>({ url: `/orgs/students/${id}`, method: 'put', data })
export const deleteStudent = (id: number) =>
  request<void>({ url: `/orgs/students/${id}`, method: 'delete' })
export const getStudentDetail = (id: number) =>
  request<Student>({ url: `/orgs/students/${id}`, method: 'get' })

// ---- 教师 ----
export const listTeachers = (params?: any) => list<Teacher>('/orgs/teachers', params)
export const createTeacher = (data: any) => request<Teacher>({ url: '/orgs/teachers', method: 'post', data })
export const updateTeacher = (id: number, data: any) =>
  request<Teacher>({ url: `/orgs/teachers/${id}`, method: 'put', data })
export const deleteTeacher = (id: number) =>
  request<void>({ url: `/orgs/teachers/${id}`, method: 'delete' })

// ---- 科目 ----
export const listSubjects = () => list<Subject>('/orgs/subjects')
export const createSubject = (data: any) => request<Subject>({ url: '/orgs/subjects', method: 'post', data })
export const updateSubject = (id: number, data: any) =>
  request<Subject>({ url: `/orgs/subjects/${id}`, method: 'put', data })
export const deleteSubject = (id: number) =>
  request<void>({ url: `/orgs/subjects/${id}`, method: 'delete' })

// ---- 任课关系 ----
export const listTcs = (params?: any) => list<TeacherClassSubject>('/orgs/teacher-class-subjects', params)
export const createTcs = (data: any) =>
  request<TeacherClassSubject>({ url: '/orgs/teacher-class-subjects', method: 'post', data })
export const deleteTcs = (id: number) =>
  request<void>({ url: `/orgs/teacher-class-subjects/${id}`, method: 'delete' })

File: backend/app/api/orgs.py
"""组织架构接口（学校 / 年级 / 班级 / 学生 / 教师 / 科目 / 任课关系）。"""

from __future__ import annotations

from flask import Blueprint, request

from ..services import org_service as svc
from ..utils.permissions import (
    admin_required,
    assert_can_access_class,
    assert_can_access_student,
    compute_data_scope,
    get_current_user,
    get_visible_class_ids,
    login_required,
    roles_required,
)
from ..utils.response import ok
from ..models import ROLE_SUPER_ADMIN

orgs_bp = Blueprint("orgs", __name__)


# ---------- 学校 ----------
@orgs_bp.get("/schools")
@login_required
def get_schools():
    user = get_current_user()
    items = svc.list_schools(keyword=request.args.get("keyword"))
    if not user.is_super and user.school_id:
        items = [i for i in items if i["id"] == user.school_id]
    return ok({"items": items, "total": len(items)})


@orgs_bp.post("/schools")
@roles_required(ROLE_SUPER_ADMIN)
def add_school():
    return ok(svc.create_school(request.get_json() or {}), "创建成功")


@orgs_bp.put("/schools/<int:school_id>")
@roles_required(ROLE_SUPER_ADMIN)
def edit_school(school_id: int):
    return ok(svc.update_school(school_id, request.get_json() or {}), "更新成功")


@orgs_bp.delete("/schools/<int:school_id>")
@roles_required(ROLE_SUPER_ADMIN)
def remove_school(school_id: int):
    svc.delete_school(school_id)
    return ok(message="已删除")


# ---------- 年级 ----------
@orgs_bp.get("/grades")
@login_required
def get_grades():
    user = get_current_user()
    school_id = request.args.get("school_id", type=int)
    if not user.is_super and user.school_id:
        school_id = user.school_id
    items = svc.list_grades(school_id=school_id)
    scope = compute_data_scope(user)
    if not scope.is_full and scope.grade_ids is not None:
        items = [i for i in items if i["id"] in scope.grade_ids]
    return ok({"items": items, "total": len(items)})


@orgs_bp.post("/grades")
@admin_required
def add_grade():
    return ok(svc.create_grade(request.get_json() or {}), "创建成功")


@orgs_bp.put("/grades/<int:grade_id>")
@admin_required
def edit_grade(grade_id: int):
    return ok(svc.update_grade(grade_id, request.get_json() or {}), "更新成功")


@orgs_bp.delete("/grades/<int:grade_id>")
@admin_required
def remove_grade(grade_id: int):
    svc.delete_grade(grade_id)
    return ok(message="已删除")


# ---------- 班级 ----------
@orgs_bp.get("/classes")
@login_required
def get_classes():
    user = get_current_user()
    grade_id = request.args.get("grade_id", type=int)
    school_id = request.args.get("school_id", type=int)
    if not user.is_super and user.school_id:
        school_id = user.school_id
    visible = get_visible_class_ids(user)
    items = svc.list_classes(grade_id=grade_id, school_id=school_id, class_ids=visible)
    return ok({"items": items, "total": len(items)})


@orgs_bp.post("/classes")
@admin_required
def add_class():
    return ok(svc.create_class(request.get_json() or {}), "创建成功")


@orgs_bp.put("/classes/<int:class_id>")
@admin_required
def edit_class(class_id: int):
    return ok(svc.update_class(class_id, request.get_json() or {}), "更新成功")


@orgs_bp.delete("/classes/<int:class_id>")
@admin_required
def remove_class(class_id: int):
    svc.delete_class(class_id)
    return ok(message="已删除")


# ---------- 学生 ----------
@orgs_bp.get("/students")
@login_required
def get_students():
    user = get_current_user()
    page = request.args.get("page", default=1, type=int)
    page_size = request.args.get("page_size", default=20, type=int)
    class_id = request.args.get("class_id", type=int)
    grade_id = request.args.get("grade_id", type=int)
    school_id = request.args.get("school_id", type=int)
    keyword = request.args.get("keyword")

    if not user.is_super and user.school_id:
        school_id = user.school_id

    visible = get_visible_class_ids(user)
    return ok(
        svc.list_students(
            class_id=class_id,
            grade_id=grade_id,
            school_id=school_id,
            keyword=keyword,
            visible_class_ids=visible,
            page=page,
            page_size=page_size,
        )
    )


@orgs_bp.get("/students/<int:student_id>")
@login_required
def get_student_detail(student_id: int):
    user = get_current_user()
    assert_can_access_student(user, student_id)
    from ..extensions import db
    from ..models import Student

    s = db.session.get(Student, student_id)
    if not s or s.is_deleted:
        from ..utils.exceptions import NotFoundError

        raise NotFoundError("学生不存在")
    return ok(svc.serialize_student(s))


@orgs_bp.post("/students")
@admin_required
def add_student():
    return ok(svc.create_student(request.get_json() or {}), "创建成功")


@orgs_bp.put("/students/<int:student_id>")
@admin_required
def edit_student(student_id: int):
    return ok(svc.update_student(student_id, request.get_json() or {}), "更新成功")


@orgs_bp.delete("/students/<int:student_id>")
@admin_required
def remove_student(student_id: int):
    svc.delete_student(student_id)
    return ok(message="已删除")


# ---------- 教师 ----------
@orgs_bp.get("/teachers")
@login_required
def get_teachers():
    user = get_current_user()
    school_id = request.args.get("school_id", type=int)
    if not user.is_super and user.school_id:
        school_id = user.school_id
    items = svc.list_teachers(school_id=school_id, keyword=request.args.get("keyword"))
    return ok({"items": items, "total": len(items)})


@orgs_bp.post("/teachers")
@admin_required
def add_teacher():
    return ok(svc.create_teacher(request.get_json() or {}), "创建成功")


@orgs_bp.put("/teachers/<int:teacher_id>")
@admin_required
def edit_teacher(teacher_id: int):
    return ok(svc.update_teacher(teacher_id, request.get_json() or {}), "更新成功")


@orgs_bp.delete("/teachers/<int:teacher_id>")
@admin_required
def remove_teacher(teacher_id: int):
    svc.delete_teacher(teacher_id)
    return ok(message="已删除")


# ---------- 科目 ----------
@orgs_bp.get("/subjects")
@login_required
def get_subjects():
    items = svc.list_subjects()
    return ok({"items": items, "total": len(items)})


@orgs_bp.post("/subjects")
@admin_required
def add_subject():
    return ok(svc.create_subject(request.get_json() or {}), "创建成功")


@orgs_bp.put("/subjects/<int:subject_id>")
@admin_required
def edit_subject(subject_id: int):
    return ok(svc.update_subject(subject_id, request.get_json() or {}), "更新成功")


@orgs_bp.delete("/subjects/<int:subject_id>")
@admin_required
def remove_subject(subject_id: int):
    svc.delete_subject(subject_id)
    return ok(message="已删除")


# ---------- 任课关系 ----------
@orgs_bp.get("/teacher-class-subjects")
@login_required
def get_tcs():
    items = svc.list_teacher_class_subjects(
        teacher_id=request.args.get("teacher_id", type=int),
        class_id=request.args.get("class_id", type=int),
        subject_id=request.args.get("subject_id", type=int),
    )
    return ok({"items": items, "total": len(items)})


@orgs_bp.post("/teacher-class-subjects")
@admin_required
def add_tcs():
    return ok(svc.create_teacher_class_subject(request.get_json() or {}), "创建成功")


@orgs_bp.delete("/teacher-class-subjects/<int:record_id>")
@admin_required
def remove_tcs(record_id: int):
    svc.delete_teacher_class_subject(record_id)
    return ok(message="已删除")


# ---------- 年级组长任职 ----------
@orgs_bp.get("/grade-heads")
@login_required
def get_grade_heads():
    items = svc.list_grade_heads(
        grade_id=request.args.get("grade_id", type=int),
        teacher_id=request.args.get("teacher_id", type=int),
    )
    return ok({"items": items, "total": len(items)})


@orgs_bp.post("/grade-heads")
@admin_required
def add_grade_head():
    payload = request.get_json() or {}
    return ok(
        svc.assign_grade_head(payload.get("teacher_id"), payload.get("grade_id")),
        "已任命",
    )


@orgs_bp.delete("/grade-heads/<int:record_id>")
@admin_required
def remove_grade_head(record_id: int):
    svc.revoke_grade_head(record_id)
    return ok(message="已撤销")
